File: medium/0005-longest-palindromic-substring/python/solution.py
from typing import Tuple


class Solution:
    def longestPalindrome(self, s: str) -> str:
        if len(s) <= 1:
            return s

        start, end = 0, 0
        for i in range(len(s)):
            # have to treat two different cases
            # to keep it simple
            # odd palindromes
            pal_odd = self.expand_centre(s, i, i)
            # even palindromes
            pal_even = self.expand_centre(s, i, i + 1)
            if pal_odd[1] - pal_odd[0] > end - start:
                start, end = pal_odd
            if pal_even[1] - pal_odd[1] > end - start:
                start, end = pal_even
        return s[start : end + 1]

    def expand_centre(self, s: str, left: int, right: int) -> Tuple[int, int]:
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        return (left + 1, right - 1)

File: medium/0005-longest-palindromic-substring/go/solution.go
package logest_palindromic_substring

func longestPalindrome(s string) string {
	max_l := 0
	max_r := 0

	for i := 0; i < len(s); i++ {
		odd_l, odd_r := expand_centre(s, i, i)
		even_l, even_r := expand_centre(s, i, i+1)
		if odd_r-odd_l > max_r-max_l {
			max_l, max_r = odd_l, odd_r
		}

		if even_r-even_l > max_r-max_l {
			max_l, max_r = even_l, even_r
		}
	}

	return s[max_l : max_r+1]
}

func expand_centre(s string, left, right int) (int, int) {
	for left >= 0 && right < len(s) && s[left] == s[right] {
		left -= 1
		right += 1
	}
	return left + 1, right - 1
}
